From 2b08878e01733938a722b283a1da15d4e5826028 Mon Sep 17 00:00:00 2001
From: Andreas Baierl <ichgeh@imkreisrum.de>
Date: Fri, 25 Nov 2022 20:28:01 +0100
Subject: [PATCH 5/7] drm: fix crtc height error

Rewrite the section with resizes the frame to fit into the given screen or
video area.

Signed-off-by: Andreas Baierl <ichgeh@imkreisrum.de>
---
 video_drm.c | 25 +++++++++++++++++--------
 1 file changed, 17 insertions(+), 8 deletions(-)

diff --git a/video_drm.c b/video_drm.c
index 66dd1ab..f421ff7 100644
--- a/video_drm.c
+++ b/video_drm.c
@@ -1312,18 +1312,27 @@ page_flip:
 	uint64_t PicWidth = DispWidth;
 	uint64_t PicHeight = DispHeight;
 
+	// resize frame to fit into video area/ screen and keep the aspect ratio
 	if (frame) {
-		PicWidth = DispHeight * av_q2d(frame->sample_aspect_ratio) *
-			   frame->width / frame->height;
-		if (!PicWidth || PicWidth > DispWidth) {
-			PicWidth = DispWidth;
-			PicHeight = PicWidth * frame->height / av_q2d(frame->sample_aspect_ratio) / frame->width;
+		double frame_sar;
+		// use frame->sample_aspect_ratio of 1.0f if undefined (0.0f), otherwise we have division by 0
+		frame_sar = av_q2d(frame->sample_aspect_ratio) ? av_q2d(frame->sample_aspect_ratio) : 1.0f;
+
+		// frame b*h < display b*h, e.g. fit a 4:3 frame into 16:9 display or area
+		if (DispWidth / DispHeight >= frame->width / frame->height) {
+			PicWidth = DispHeight * frame->width / frame->height * frame_sar;
+			if (PicWidth <= 0 || PicWidth > DispWidth) {
+				PicWidth = DispWidth;
+			}
+		// frame b*h >= display b*h, e.g. fit a 16:9 frame into 4:3 display or area
+		} else {
+			PicHeight = DispWidth * frame->height / frame->width / frame_sar;
+			if (PicHeight <= 0 || PicHeight > DispHeight) {
+				PicHeight = DispHeight;
+			}
 		}
 	}
 
-	if (!PicWidth || PicWidth > DispWidth)
-		PicWidth = DispWidth;
-
 	SetPlane(ModeReq, render->planes[VIDEO_PLANE]->plane_id, render->crtc_id, buf->fb_id,
 		 DispX + (DispWidth - PicWidth) / 2, DispY + (DispHeight - PicHeight) / 2, PicWidth, PicHeight,
 		 0, 0, buf->width, buf->height);
-- 
2.30.2

